File: EyesRightMac/Sources/EyesRightMac/Core/PoseDetector.swift
import CoreGraphics
import Foundation
import OnnxRuntimeBindings

final class PoseDetector: @unchecked Sendable {
    private let session: ORTSession

    init() throws {
        guard let modelURL = AppResources.url(forResource: "pet_eye_best", withExtension: "onnx") else {
            throw PipelineError.modelNotFound
        }
        let env = try ORTEnv(loggingLevel: .warning)
        let options = try ORTSessionOptions()
        session = try ORTSession(env: env, modelPath: modelURL.path, sessionOptions: options)
    }

    func detect(in image: CGImage, confThreshold: Float = OverlayConstants.confThreshold) throws -> [EyePair] {
        let letterbox = try ImageProcessor.letterbox(image)
        let inputShape: [NSNumber] = [1, 3, 640, 640]
        let inputData = Data(bytes: letterbox.tensor, count: letterbox.tensor.count * MemoryLayout<Float>.size)
        let inputTensor = try ORTValue(
            tensorData: NSMutableData(data: inputData),
            elementType: .float,
            shape: inputShape
        )

        let outputs = try session.run(
            withInputs: ["images": inputTensor],
            outputNames: ["output0"],
            runOptions: nil
        )

        guard let outputTensor = outputs["output0"],
              let outputData = try outputTensor.tensorData() as Data?
        else {
            throw PipelineError.inferenceFailed
        }

        let count = 8400
        let floats = outputData.withUnsafeBytes { rawBuffer in
            Array(rawBuffer.bindMemory(to: Float.self))
        }

        var candidates: [(score: Float, cx: Float, cy: Float, w: Float, kpts: [(CGPoint, Float)])] = []

        for index in 0..<count {
            func value(_ channel: Int) -> Float {
                floats[channel * count + index]
            }

            let score = value(4)
            guard score >= confThreshold else { continue }

            let cx = value(0)
            let cy = value(1)
            let boxW = value(2)

            var keypoints: [(CGPoint, Float)] = []
            for kptIndex in 0..<3 {
                let channel = 5 + kptIndex * 3
                let point = CGPoint(x: CGFloat(value(channel)), y: CGFloat(value(channel + 1)))
                let mapped = ImageProcessor.mapPointToOriginal(point, letterbox: letterbox)
                keypoints.append((mapped, value(channel + 2)))
            }

            candidates.append((score, cx, cy, boxW, keypoints))
        }

        candidates.sort { $0.score > $1.score }
        guard let best = candidates.first else { return [] }

        var left = best.kpts[0].0
        var right = best.kpts[1].0
        if left.x > right.x {
            swap(&left, &right)
        }

        let interEye = hypot(right.x - left.x, right.y - left.y)
        let mid = CGPoint(x: (left.x + right.x) / 2, y: (left.y + right.y) / 2)
        // 两眼连线的法向（图像顶左，y 向下）；模型缺鼻点时给一个「额头侧」先验供 NoseRefiner 反射
        let nx = -(right.y - left.y) / max(interEye, 1e-6)
        let ny = (right.x - left.x) / max(interEye, 1e-6)
        let foreheadCue = CGPoint(
            x: mid.x - nx * interEye * 0.45,
            y: mid.y - ny * interEye * 0.45
        )
        // 始终保留原始 kpt2（即便偏到额头）；由 NoseRefiner 校正到鼻头
        let nose = (best.kpts.count > 2 && best.kpts[2].1 >= OverlayConstants.noseConfThreshold)
            ? best.kpts[2].0
            : foreheadCue

        let kptConf = min(best.kpts[0].1, best.kpts[1].1)
        let boxWidth = CGFloat(best.w / letterbox.scale)

        return [
            EyePair(
                left: left,
                right: right,
                nose: nose,
                confidence: min(best.score, kptConf),
                boxWidth: boxWidth
            ),
        ]
    }
}

File: EyesRightFlutter/lib/inference/pose_detector.dart
import 'dart:math' as math;
import 'dart:typed_data';
import 'dart:ui' as ui;

import 'package:flutter_onnxruntime/flutter_onnxruntime.dart';

import '../models/eye_models.dart';

class PoseDetector {
  PoseDetector._(this._session);

  final OrtSession _session;
  static const _modelAsset = 'assets/models/pet_eye_best.onnx';
  static const _anchors = 8400;

  static Future<PoseDetector> create() async {
    final ort = OnnxRuntime();
    final session = await ort.createSessionFromAsset(_modelAsset);
    return PoseDetector._(session);
  }

  Future<List<EyePair>> detect(LetterboxResult letterbox, {double confThreshold = OverlayConstants.confThreshold}) async {
    final input = await OrtValue.fromList(
      letterbox.tensor,
      const [1, 3, OverlayConstants.inputSize, OverlayConstants.inputSize],
    );
    try {
      final outputs = await _session.run({'images': input});
      final output = outputs.values.first;
      try {
        final flat = await output.asFlattenedList();
        final floats = Float32List(flat.length);
        for (var i = 0; i < flat.length; i++) {
          floats[i] = (flat[i] as num).toDouble();
        }
        return _parse(floats, letterbox, confThreshold);
      } finally {
        await output.dispose();
      }
    } finally {
      await input.dispose();
    }
  }

  List<EyePair> _parse(Float32List floats, LetterboxResult letterbox, double confThreshold) {
    double value(int channel, int index) => floats[channel * _anchors + index];

    var bestScore = -1.0;
    var bestW = 0.0;
    final bestKpts = <(ui.Offset, double)>[];

    for (var index = 0; index < _anchors; index++) {
      final score = value(4, index);
      if (score < confThreshold || score <= bestScore) continue;

      bestScore = score;
      bestW = value(2, index);
      bestKpts
        ..clear()
        ..addAll(
          List.generate(3, (k) {
            final channel = 5 + k * 3;
            final mapped = letterbox.mapToOriginal(
              ui.Offset(value(channel, index), value(channel + 1, index)),
            );
            return (mapped, value(channel + 2, index));
          }),
        );
    }

    if (bestScore < 0 || bestKpts.length < 2) {
      return const [];
    }

    var left = bestKpts[0].$1;
    var right = bestKpts[1].$1;
    if (left.dx > right.dx) {
      final tmp = left;
      left = right;
      right = tmp;
    }

    final kptConf = bestKpts[0].$2 < bestKpts[1].$2 ? bestKpts[0].$2 : bestKpts[1].$2;
    final interEye = hypot(right.dx - left.dx, right.dy - left.dy);
    final mid = ui.Offset(
      (left.dx + right.dx) / 2,
      (left.dy + right.dy) / 2,
    );
    final nx = -(right.dy - left.dy) / math.max(interEye, 1e-6);
    final ny = (right.dx - left.dx) / math.max(interEye, 1e-6);
    // 缺鼻点时给额头侧先验，供 NoseRefiner 反射到鼻头
    final foreheadCue = ui.Offset(
      mid.dx - nx * interEye * 0.45,
      mid.dy - ny * interEye * 0.45,
    );
    final noseConf = bestKpts.length > 2 ? bestKpts[2].$2 : 0.0;
    final nose = noseConf >= OverlayConstants.noseConfThreshold
        ? bestKpts[2].$1
        : foreheadCue;

    return [
      EyePair(
        left: left,
        right: right,
        nose: nose,
        confidence: bestScore < kptConf ? bestScore : kptConf,
        boxWidth: bestW / letterbox.scale,
      ),
    ];
  }

  Future<void> close() => _session.close();
}
